File: src/main/java/discretemath/bool/expression/forms/Maxterm.java
package discretemath.bool.expression.forms;

import java.util.List;

import discretemath.bool.expression.BooleanExpression;
import discretemath.bool.operator.FunctionallyCompleteOperatorsSet;

public interface Maxterm extends BooleanExpression {

	List<Literal> literals();

	@Override
	default BooleanExpression expressViaOperators(FunctionallyCompleteOperatorsSet operatorsSet) {
		throw new UnsupportedOperationException("Not implemented yet");
	}

	@Override
	default BooleanExpression normalize() {
		return this;
	}
}

File: src/main/scala/discretemath/bool/expression/forms/Maxterm.scala
package discretemath.bool.expression.forms;

import discretemath.bool.expression.BooleanExpression
import discretemath.bool.expression.atomic.BooleanVariable
import discretemath.bool.expression.exception.VariableValueNotSpecifiedException
import discretemath.bool.operator.FunctionallyCompleteOperatorsSet
import discretemath.bool.operator.Operators.OR

import scala.collection.immutable.ListSet;

case class Maxterm(literals: ListSet[Literal]) extends BooleanExpression {

  override def getVariables: Set[BooleanVariable] = literals.map(_.variable)

  @throws[VariableValueNotSpecifiedException]
  override def evaluate(arguments: Map[BooleanVariable, Boolean]): Boolean = {
    val values: Seq[Boolean] = literals.view.map(literal => literal.evaluate(arguments)).toSeq

    OR.evaluate(values)
  }

  override def expressViaOperators(operatorsSet: FunctionallyCompleteOperatorsSet): BooleanExpression = OR.injectUsing(operatorsSet, literals.toSeq)

  override def minimize: BooleanExpression = this

  override def isSatisfiable: Boolean = true

  override def toString: String = literals.view.map(_.toString).mkString("+")
}